pybricks/tools/pb_module_tools.c

```diff
@@ -152,22 +152,52 @@ mp_obj_t pb_module_tools_pbio_task_wait_or_await(pbio_task_t *task) {
 }
 
 /**
- * Reads one byte from stdin without blocking.
+ * Reads one byte or character from stdin without blocking.
  *
- * @returns The integer value of the byte read or @c None if no data is available.
+ * @param [in]  chr     Choose @c True to return a printable character or @c False to return an integer.
+ * @param [in]  last    Choose @c True to read until the last byte is read or @c False to return the first byte.
+ *
+ * @returns The value read or @c None if no valid data is available.
  */
-STATIC mp_obj_t pb_module_tools_read_input_byte(void) {
-    if (!(mp_hal_stdio_poll(MP_STREAM_POLL_RD) & MP_STREAM_POLL_RD)) {
-        // No bytes available.
+STATIC mp_obj_t pb_module_tools_read_input_byte(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    PB_PARSE_ARGS_FUNCTION(n_args, pos_args, kw_args,
+        PB_ARG_DEFAULT_FALSE(chr),
+        PB_ARG_DEFAULT_FALSE(last));
+
+    int chr = -1;
+
+    while ((mp_hal_stdio_poll(MP_STREAM_POLL_RD) & MP_STREAM_POLL_RD)) {
+        // REVISIT: In theory, this should not block if mp_hal_stdio_poll() and
+        // mp_hal_stdin_rx_chr() are implemented correctly and nothing happens
+        // in a thread/interrupt/kernel that changes the state.
+        chr = mp_hal_stdin_rx_chr();
+
+        // For last=True, read up to last, otherwise return first byte.
+        if (!mp_obj_is_true(last_in)) {
+            break;
+        }
+    }
```

Comment:
Should we be skipping non-printable ASCII characters in the loop when `chr=True`?

Reply:
Good point. I've changed the argument order to make it clearer we first care about the first/last _byte_, and then look at converting to a single character string, if `chr` is `True`.

Reply:
My question is if `last=False, chr=True`, and the buffer contains 3 non-printable characters and one printable character, is the desired behavoir to return `None` 3 times and then return the printable character? Because it seems like that's what would happen now.